pkg/secrets/vault.go

```diff
@@ -2,23 +2,222 @@ package secrets
+	Url        string `json:"url" yaml:"url"`
+	AuthMethod string `json:"authMethod" yaml:"authMethod"`
+	Role       string `json:"role" yaml:"role"`
+	Path       string `json:"path" yaml:"path"`
+	Token      string
+}
+
 type VaultSecret struct {
 	engine        string
-	namespace     string
+	path          string
 	key           string
-	base64Encoded bool
+	base64Encoded string
 }
 
 type VaultDecrypter struct {
-	secretConfig string
+	encryptedSecret string
 }
 
-func NewVaultDecrypter(secretConfig string) *VaultDecrypter {
-	return &VaultDecrypter{secretConfig}
+func NewVaultDecrypter(encryptedSecret string) *VaultDecrypter {
+	return &VaultDecrypter{encryptedSecret}
 }
 
 func (v *VaultDecrypter) Decrypt() (string, error) {
-	return "", fmt.Errorf("vault secrets unsupported")
+	if (VaultConfig{}) == Registry.VaultConfig {
+		return "", fmt.Errorf("error: vault secrets configuration not found")
+	}
+	vaultSecret, err := ParseVaultEncryptedSecret(v.encryptedSecret)
+	if err != nil {
+		return "", fmt.Errorf("error parsing vault secret syntax - %s", err)
+	}
+
+	if Registry.VaultConfig.Token == "" {
+		token, err := v.FetchVaultToken()
+		if err != nil {
+			return "", fmt.Errorf("error fetching vault token - %s", err)
+		}
+		Registry.VaultConfig.Token = token
+	}
+
+	secret, err := v.FetchSecret(vaultSecret)
+	if err != nil && strings.Contains(err.Error(), "403") {
+		// get new token and retry in case our saved token is no longer valid
+		return v.RetryFetchSecret(vaultSecret)
+	}
+	return secret, nil
+}
+
+func ValidateVaultConfig(vaultConfig VaultConfig) error {
+	if (VaultConfig{}) == vaultConfig {
+		return fmt.Errorf("vault secrets not configured in service profile yaml")
+	}
+	if vaultConfig.Enabled == false {
+		return fmt.Errorf("vault secrets disabled")
+	}
+	if vaultConfig.Url == "" {
+		return fmt.Errorf("vault url required")
+	}
+	if vaultConfig.AuthMethod == "" {
+		return fmt.Errorf("auth method required")
+	}
+
+	if vaultConfig.AuthMethod == "TOKEN" {
+		if token := os.Getenv("VAULT_TOKEN"); token == "" {
+			return fmt.Errorf("VAULT_TOKEN environment variable not set")
+		}
+	} else if vaultConfig.AuthMethod == "KUBERNETES" {
+		if vaultConfig.Path == "" || vaultConfig.Role == "" {
+			return fmt.Errorf("path and role both required for Kubernetes auth method")
+		}
+	} else {
+		return fmt.Errorf("unknown Vault secrets auth method: %q", vaultConfig.AuthMethod)
+	}
+
+	return nil
+}
+
+func ParseVaultEncryptedSecret(encryptedSecret string) (VaultSecret, error) {
+	var vaultSecret VaultSecret
+	configs := strings.Split(encryptedSecret, "!")
+	if len(configs) < 2 {
+		return VaultSecret{}, fmt.Errorf("illegal format: %q", encryptedSecret)
+	}
+	for _, element := range configs {
+		kv := strings.Split(element, ":")
+		if len(kv) < 2 {
+			return VaultSecret{}, fmt.Errorf("illegal format for key-value pair in %q: %s",
+				encryptedSecret, element)
+		}
+		switch kv[0] {
+		case "encrypted":
+			// do nothing
```

Comment:
that's a bizarre parsing. Shouldn't it be already parsed before you determine the engine?

Reply:
Is this tied to the test case for "plainTextSecret"?  Where there are "secrets" that aren't encrypted?